framework/sandstone_p.h

```diff
@@ -320,6 +324,67 @@ class LoggingStream
     friend LoggingStream logging_user_messages_stream(int thread_num, int level);
 };
 
+class DeviceSchedule {
+public:
+    virtual int get_next_cpu() = 0;
+    virtual ~DeviceSchedule() = default;
+};
+
+class QueueDeviceSchedule : public DeviceSchedule {
+public:
+    QueueDeviceSchedule(struct cpu_info *cpu_info, int q_size)
+    {
+        q_idx.store(0, std::memory_order_relaxed);
+        queue.reserve(q_size);
+
+        // Populate queue
+        for (int i=0; i<q_size; i++)
+            queue.push_back(cpu_info[i].cpu_number);
+    }
+
+    int get_next_cpu() override
+    {
+        // Returns next CPU from the queue
+        std::lock_guard lock(q_mutex);
+        int idx = q_idx.fetch_add(1, std::memory_order_relaxed);
+        if (q_idx.load() >= queue.size())
+            shuffle_queue();
+        return queue[idx];
+    }
+
+private:
+    std::atomic<int> q_idx;
+    std::vector<int> queue;
+    std::mutex q_mutex;
+
+    void shuffle_queue() {
+        // Must be called with mutex locked
+        std::default_random_engine rng(random32());
+        std::shuffle(queue.begin(), queue.end(), rng);
+        q_idx.store(0, std::memory_order_relaxed);
+    }
+};
+
+class RandomDeviceSchedule : public DeviceSchedule {
+public:
+    RandomDeviceSchedule(struct cpu_info *cpu_info, int q_size)
+    {
+        queue.reserve(q_size);
+
+        // Populate queue
+        for (int i=0; i<q_size; i++)
+            queue.push_back(cpu_info[i].cpu_number);
```

Comment:
Same comment here, this limits the scope of CPUs that we migrate between. If we run 4 software threads on 64 hw threads, they will always migrate between the first 4 hw threads, the rest will be always idle. In case of random, it should be populated with all available CPUs (to this slice).

Reply:
I think this will be pending for a future PR.